File: sdk/go/run_helpers.go
package macrofold

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// RunFailedError distinguishes an unsuccessful run from an HTTP/transport failure.
type RunFailedError struct {
	RunID       string
	Status      string
	FailureCode string
	Result      *RunResult
}

func (e *RunFailedError) Error() string {
	return fmt.Sprintf("run %s ended with status %s (persistence: %s)", e.RunID, e.Status, e.Result.PersistenceStatus)
}

type WaitTimeoutError struct{ RunID string }

func (e *WaitTimeoutError) Error() string {
	return fmt.Sprintf("timed out waiting for run %s; the agent has not been cancelled", e.RunID)
}
func (e *WaitTimeoutError) Unwrap() error { return context.DeadlineExceeded }

// Events exposes structured events; cancelling the context detaches without cancelling execution.
func (r *RunsResource) Events(ctx context.Context, runID, after string, receive func(Event) error, options ...RequestOption) error {
	return r.Stream(ctx, runID, after, receive, options...)
}

// StreamText delivers assistant text only, then verifies the final execution/persistence outcome.
func (r *RunsResource) StreamText(ctx context.Context, runID, after string, receive func(string) error, options ...RequestOption) error {
	err := r.Events(ctx, runID, after, func(event Event) error {
		if text, ok := event.Data["text"].(string); event.Type == "output.delta" && ok && text != "" {
			return receive(text)
		}
		return nil
	}, options...)
	if err != nil {
		return err
	}
	_, err = r.Wait(ctx, runID, options...)
	return err
}

// Wait returns the durable result. Use a context deadline to stop waiting, not the remote agent.
func (r *RunsResource) Wait(ctx context.Context, runID string, options ...RequestOption) (result *RunResult, err error) {
	defer func() {
		if errors.Is(err, context.DeadlineExceeded) && ctx.Err() == context.DeadlineExceeded {
			err = &WaitTimeoutError{RunID: runID}
		}
	}()
	for {
		if err = ctx.Err(); err != nil {
			return nil, err
		}
		run, err := r.Get(ctx, runID, options...)
		if err != nil {
			return nil, err
		}
		switch run.Status {
		case "succeeded", "failed", "cancelled", "timed_out":
			result, err := r.GetResult(ctx, runID, options...)
			if err != nil {
				return nil, err
			}
			if err := ctx.Err(); err != nil {
				return nil, err
			}
			if result.Final && result.PersistenceStatus != "pending" {
				if run.Status != "succeeded" || result.ExecutionOutcome != "success" || (result.PersistenceStatus != "verified" && result.PersistenceStatus != "not_required") {
					return nil, &RunFailedError{RunID: runID, Status: run.Status, FailureCode: run.GetFailureCode(), Result: result}
				}
				return result, nil
			}
		}
		timer := time.NewTimer(time.Second)
		select {
		case <-ctx.Done():
			timer.Stop()
			return nil, ctx.Err()
		case <-timer.C:
		}
	}
}

File: sdk/rust/src/run_helpers.rs
use crate::{
    models::{Event, RunResult},
    resources::RunsResource,
    ClientError,
};
use std::time::Duration;

#[derive(Debug)]
pub struct RunFailedError {
    pub run_id: String,
    pub status: crate::models::run::Status,
    pub failure_code: Option<String>,
    pub result: RunResult,
}
impl std::fmt::Display for RunFailedError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Run {} ended with status {:?} (persistence: {}).",
            self.run_id, self.status, self.result.persistence_status
        )
    }
}
impl std::error::Error for RunFailedError {}
#[derive(Debug)]
pub struct WaitTimeoutError {
    pub run_id: String,
}
impl std::fmt::Display for WaitTimeoutError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Timed out waiting for run {}. The agent has not been cancelled.",
            self.run_id
        )
    }
}
impl std::error::Error for WaitTimeoutError {}

impl RunsResource<'_> {
    pub async fn events(
        &self,
        run_id: &str,
        after: &str,
        receive: impl FnMut(Event) -> bool,
    ) -> Result<(), ClientError> {
        self.stream(run_id, after, receive).await
    }
    pub async fn stream_text(
        &self,
        run_id: &str,
        after: &str,
        mut receive: impl FnMut(String) -> bool,
    ) -> Result<(), ClientError> {
        let mut detached = false;
        self.events(run_id, after, |event| {
            if event.r#type == "output.delta" {
                if let Some(text) = event
                    .data
                    .get("text")
                    .and_then(|value| value.as_str())
                    .filter(|text| !text.is_empty())
                {
                    detached = !receive(text.to_owned());
                }
            }
            !detached
        })
        .await?;
        if !detached {
            self.wait(run_id).await?;
        }
        Ok(())
    }
    /// Dropping this future stops waiting; it never cancels the remote run.
    pub async fn wait(&self, run_id: &str) -> Result<RunResult, ClientError> {
        use crate::models::run::Status;
        loop {
            let run = self.get(run_id).await?;
            if matches!(
                run.status,
                Status::Succeeded | Status::Failed | Status::Cancelled | Status::TimedOut
            ) {
                let result = self.get_result(run_id).await?;
                if result.r#final && result.persistence_status != "pending" {
                    if run.status != Status::Succeeded
                        || result.execution_outcome != "success"
                        || !matches!(
                            result.persistence_status.as_str(),
                            "verified" | "not_required"
                        )
                    {
                        return Err(Box::new(RunFailedError {
                            run_id: run_id.into(),
                            status: run.status,
                            failure_code: run.failure_code,
                            result,
                        }));
                    }
                    return Ok(result);
                }
            }
            tokio::time::sleep(Duration::from_secs(1)).await;
        }
    }
    pub async fn wait_with_timeout(
        &self,
        run_id: &str,
        timeout: Duration,
    ) -> Result<RunResult, ClientError> {
        if timeout.is_zero() {
            return Err(Box::new(WaitTimeoutError {
                run_id: run_id.into(),
            }));
        }
        tokio::time::timeout(timeout, self.wait(run_id))
            .await
            .map_err(|_| -> ClientError {
                Box::new(WaitTimeoutError {
                    run_id: run_id.into(),
                })
            })?
    }
}
